goneb_services_test.go

```diff
@@ -54,3 +57,310 @@ func TestConfigureClient(t *testing.T) {
+
+	// send neb an invite to a room
+	reqChan <- `{
+		"next_batch":"11_22_33_44",
+		"rooms": {
+			"invite": {
+				"!greatdekutree:hyrule": {"invite_state": {"events": [{
+					"type": "m.room.member",
+					"sender": "@navi:hyrule",
+					"content": {"membership": "invite"},
+					"state_key": "@link:hyrule",
+					"origin_server_ts": 10000,
+					"unsigned": {"age": 100},
+					"event_id": "evt123"
+				}]}}}
+		}
+	}`
+
+	// wait for it to be processed
+	reqChan <- `{"next_batch":"11_22_33_44", "rooms": {}}`
+
+	expectedRoom := "%21greatdekutree:hyrule"
+	if joinedRoom != expectedRoom {
+		t.Errorf("Expected join for room %v, got %v", expectedRoom, joinedRoom)
+	}
+	if expectedBody := `{"inviter":"@navi:hyrule"}`; string(joinedRoomBody) != expectedBody {
+		t.Errorf("Expected join message body to be %v, got %v", expectedBody, string(joinedRoomBody))
+	}
+
+	// send neb the room state: encrypted with one member
+	reqChan <- `{
+		"next_batch":"11_22_33_44",
+		"rooms": {
+			"join": {"!greatdekutree:hyrule": {"timeline": {"events": [{
+					"type": "m.room.encryption",
+					"state_key": "",
+					"content": {
+						"algorithm": "m.megolm.v1.aes-sha2"
+					}
+				}, {
+					"type": "m.room.member",
+					"sender": "@navi:hyrule",
+					"content": {"membership": "join", "displayname": "Navi"},
+					"state_key": "@navi:hyrule",
+					"origin_server_ts": 100,
+					"event_id": "evt124"
+				}]}}}
+		}
+	}`
+
+	// wait for it to be processed
+	reqChan <- `{"next_batch":"11_22_33_44", "rooms": {}}`
+
+	// DB is initialized, store the megolm sessions from before for the bot to be able to decrypt and encrypt
+	sqlDB, dialect := database.GetServiceDB().(*database.ServiceDB).GetSQLDb()
+	cryptoStore := crypto.NewSQLCryptoStore(sqlDB, dialect, "mastersword", []byte("masterswordpickle"), clients.CryptoMachineLogger{})
+	if err := cryptoStore.AddOutboundGroupSession(ogsBot); err != nil {
+		t.Errorf("Error storing bot OGS: %v", err)
+	}
+
+	if err := cryptoStore.PutGroupSession("!greatdekutree:hyrule", identityKeyMock, igsBot.ID(), igsBot); err != nil {
+		t.Errorf("Error storing bot IGS: %v", err)
+	}
+
+	plaintext := `{"room_id":"!greatdekutree:hyrule","type":"m.room.message","content":{"body":"!echo save zelda","msgtype":"m.text"}}`
+	ciphertext, err := ogsMock.Encrypt([]byte(plaintext))
+	if err != nil {
+		t.Errorf("Error encrypting bytes: %v", err)
+	}
+
+	// send neb an !echo message, encrypted with our mock OGS which it has an IGS for
+	reqChan <- fmt.Sprintf(`{
+		"next_batch":"11_22_33_44",
+		"rooms": {
+			"join": {"!greatdekutree:hyrule": {"timeline": {"events": [{
+					"type": "m.room.encrypted",
+					"sender": "@navi:hyrule",
+					"content": {
+						"algorithm":"m.megolm.v1.aes-sha2",
+						"sender_key":"%s",
+						"ciphertext":"%s",
+						"session_id":"%s"
+					},
+					"origin_server_ts": 10000,
+					"unsigned": {"age": 100},
+					"event_id": "evt125"
+				}]}}}
+		}
+	}`, identityKeyMock, string(ciphertext), ogsMock.ID())
```

Comment:
Careful doing this as a JSON string, are we sure these 3 values only ever contain alphanumerics (no newlines or special chars?)

Reply:
Yes, they're all base64.